File: backend/crm/subscription_dates.py
from datetime import timedelta

from .group_schedule import DAY_TO_WEEKDAY, normalize_schedule_days


def _end_date_by_days(start_date, lessons_count, schedule_days):
    days = normalize_schedule_days(schedule_days)
    weekdays = {DAY_TO_WEEKDAY[day] for day in days}
    if not weekdays or lessons_count <= 0:
        return None

    current = start_date
    matched = 0
    guard = 0
    while guard < 370:
        if current.weekday() in weekdays:
            matched += 1
            if matched == lessons_count:
                return current
        current += timedelta(days=1)
        guard += 1
    return None


def calculate_subscription_end_date(start_date, lessons_count=None, validity_days=None, group=None, service_schedule_days=None):
    if not start_date:
        return None

    lessons_count = int(lessons_count or 0)
    validity_days = int(validity_days or 0)

    if group and lessons_count > 0:
        end_date = _end_date_by_days(start_date, lessons_count, getattr(group, 'schedule_days', []))
        if end_date:
            return end_date

    if lessons_count > 0:
        end_date = _end_date_by_days(start_date, lessons_count, service_schedule_days or [])
        if end_date:
            return end_date

    if validity_days > 0:
        return start_date + timedelta(days=validity_days - 1)

    if lessons_count == 4:
        return start_date + timedelta(days=27)
    if lessons_count == 8:
        return start_date + timedelta(days=30)

    return None

File: frontend/src/utils/subscriptionDates.js
const dayToWeekday = {
  monday: 1,
  tuesday: 2,
  wednesday: 3,
  thursday: 4,
  friday: 5,
  saturday: 6,
  sunday: 0,
};

export const weekdayLabels = {
  monday: 'ПН',
  tuesday: 'ВТ',
  wednesday: 'СР',
  thursday: 'ЧТ',
  friday: 'ПТ',
  saturday: 'СБ',
  sunday: 'ВС',
};

export const weekdayOptions = Object.entries(weekdayLabels).map(([value, label]) => ({ value, label }));

const toDate = (value) => {
  if (!value) return null;
  const date = new Date(`${value}T00:00:00`);
  return Number.isNaN(date.getTime()) ? null : date;
};

const toIso = (date) => date.toISOString().slice(0, 10);

export function formatScheduleDays(days = []) {
  return (Array.isArray(days) ? days : []).map((day) => weekdayLabels[day]).filter(Boolean).join(', ');
}

export function calculateEndDateByWeekdays(startDate, lessonsCount, scheduleDays = []) {
  const date = toDate(startDate);
  const total = Number(lessonsCount || 0);
  const weekdays = new Set((Array.isArray(scheduleDays) ? scheduleDays : []).map((day) => dayToWeekday[day]).filter((day) => day !== undefined));
  if (!date || total <= 0 || !weekdays.size) return '';

  let matched = 0;
  let guard = 0;
  while (guard < 370) {
    if (weekdays.has(date.getDay())) {
      matched += 1;
      if (matched === total) return toIso(date);
    }
    date.setDate(date.getDate() + 1);
    guard += 1;
  }
  return '';
}

export function addValidityDays(startDate, days) {
  const date = toDate(startDate);
  const total = Number(days || 0);
  if (!date || total <= 0) return '';
  date.setDate(date.getDate() + total - 1);
  return toIso(date);
}

export function calculateEndDateFromService(startDate, service, groupScheduleDays = []) {
  const lessonsCount = service?.lessons_count || 0;
  return (
    calculateEndDateByWeekdays(startDate, lessonsCount, groupScheduleDays)
    || calculateEndDateByWeekdays(startDate, lessonsCount, service?.schedule_days)
    || addValidityDays(startDate, service?.validity_days)
    || addValidityDays(startDate, lessonsCount === 4 ? 28 : lessonsCount === 8 ? 31 : 0)
  );
}
